### PR TITLE
[FLINK-9194][history] Add HistoryServer support to flip6

### PR DESCRIPTION
**Pull request by zentol**
This PR subsumes #5888.

## What is the purpose of the change

This PR adds HistoryServer support to Flip6, i.e. allows the Dispatcher to archive finished jobs. Additionally the HistoryServer was updated to reflect the latest changes to the job overview page.

## Brief change log

* fix job overview merging of the HistoryServer (#5888)
* replace some manual JSON reading/writing with mappings to/from `MultipleJobDetails`
* add a convenience `ArchivedJson` constructor that accepts a `ResponseBody`
* add `JsonArchivist` argument to `Dispatcher` constructor
  * the archivist creates a `Collection<ArchivedJson>` which is passed to `FsJobArchivist` for writing to disk
  * `WebMonitorEndpoint` now implements `JsonArchivist`
* Modify relevant handlers to implement `JsonArchivist`
  * this is different to the existing implementation of using static factories
  * archiving handlers are detected by the `WebMonitorEndpoint` which is significantly less brittle as the existing approach of separately listing archivers
* rework the `HistoryServerTest` to run multiple small jobs against a `MiniClusterResource`
  * this better reflects the actual behavior and ensures we don't break things again

## Verifying this change

* run `HistoryServerTest`
* manually verify that Flip6 clusters can archive jobs, and ensure that HistoryServer can read them.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java [hunk 1]
```diff
@@ -211,7 +215,7 @@ public void run() {
 						}
 					}
 					if (updateOverview) {
-						updateJobOverview(webDir);
+						updateJobOverview(webOverviewDir, webDir);
 					}
```
Comment (yuqi1129) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: Could not figure out the function of `numFinishedPolls.countDown();`in line 224, seems that `numFinishedPolls`   is useless as it was initialized 0

Comment (yuqi1129) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: In test case, it's quite clear this is make the HS has polled file before read, That is to say adding `numFinishedPolls` is just for test purpose, am i right ？

Comment (zentol) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: yes.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java [hunk 2]
```diff
@@ -162,7 +166,7 @@ public void run() {
 									String json = archive.getJson();
 
 									File target;
-									if (path.equals("/joboverview")) {
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: this change effectively kills backwards compatibility; i.e. a 1.4 HS cannot properly handle 1.5 archives, and a 1.5 HS cannot properly handle 1.4 archives. What is missing is a compatibility check for the old URL.

Comment (yuqi1129) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java]: Agree

**Pull request closed without merging**